cupyx/scipy/sparse/_index.py

```diff
@@ -0,0 +1,388 @@
+        Bx = Bx[::-1].copy()
+
+    return Bj, Bx
+
+
+class IndexMixin(object):
+    """
+    This class provides common dispatching and validation logic for indexing.
+    """
+
+    def __getitem__(self, key):
+        row, col = self._parse_indices(key)
+        # Dispatch to specialized methods.
+        if isinstance(row, _int_scalar_types):
+            if isinstance(col, _int_scalar_types):
+                return self._get_intXint(row, col)
+            elif isinstance(col, slice):
+                return self._get_intXslice(row, col)
+            elif col.ndim == 1:
+                return self._get_intXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif isinstance(row, slice):
+            if isinstance(col, _int_scalar_types):
+                return self._get_sliceXint(row, col)
+            elif isinstance(col, slice):
+                if row == slice(None) and row == col:
+                    return self.copy()
+                return self._get_sliceXslice(row, col)
+            elif col.ndim == 1:
+                return self._get_sliceXarray(row, col)
+            raise IndexError('index results in >2 dimensions')
+        elif row.ndim == 1:
+            if isinstance(col, _int_scalar_types):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                return self._get_arrayXslice(row, col)
+        else:  # row.ndim == 2
+            if isinstance(col, _int_scalar_types):
+                return self._get_arrayXint(row, col)
+            elif isinstance(col, slice):
+                raise IndexError('index results in >2 dimensions')
+            elif row.shape[1] == 1 and (col.ndim == 1 or col.shape[0] == 1):
+                # special case for outer indexing
+                return self._get_columnXarray(row[:, 0], col.ravel())
+
+        # The only remaining case is inner (fancy) indexing
+        row, col = cupy.broadcast_arrays(row, col)
+        if row.shape != col.shape:
+            raise IndexError('number of row and column indices differ')
+        if row.size == 0:
+            return self.__class__(cupy.atleast_2d(row).shape, dtype=self.dtype)
+        return self._get_arrayXarray(row, col)
+
+    def _parse_indices(self, key):
+        M, N = self.shape
+        row, col = _unpack_index(key)
+
+        # Scipy calls sputils.isintlike() rather than
+        # isinstance(x, _int_scalar_types). Comparing directly to int
+        # here to minimize the impact of nested exception catching
+
+        if isinstance(row, _int_scalar_types):
+            row = _normalize_index(row, M, 'row')
+        elif not isinstance(row, slice):
+            row = self._asindices(row, M)
+
+        if isinstance(col, _int_scalar_types):
+            col = _normalize_index(col, N, 'column')
+        elif not isinstance(col, slice):
+            col = self._asindices(col, N)
+
+        return row, col
+
+    def _asindices(self, idx, length):
```

Comment:
Why not support for integer scalar and slice case?

Reply:
The impact of this function has been minimized in this PR since we removed the indexing, but this function is specifically for indexing cases.